lte/gateway/c/oai/lib/pipelined_client/PipelinedServiceClient.cpp

```diff
@@ -0,0 +1,713 @@
+  sid->set_type(SubscriberID::IMSI);
+
+  // Set the precedence
+  request.set_precedence(flow_precedence);
+
+  // Set the APN
+  request.set_apn(apn);
+
+  // Set the ue_state
+  config_ue_session_state(ue_state, request);
+
+  auto local_response = new AsyncLocalResponse<UESessionContextResponse>(
+      std::move(callback), RESPONSE_TIMEOUT);
+
+  auto response_reader = client.stub_->AsyncUpdateUEState(
+      local_response->get_context(), request, &client.queue_);
+  local_response->set_response_reader(std::move(response_reader));
+  return RETURNok;
+}
+
+//                    ADD : v4v6 with flow dl
+//-------------------------------------------------------------------------
+int PipelinedServiceClient::UpdateUEIPv4v6SessionSetWithFlowdl(
+    const struct in_addr& ue_ipv4_addr, struct in6_addr& ue_ipv6_addr, int vlan,
+    struct in_addr& enb_ipv4_addr, uint32_t in_teid, uint32_t out_teid,
+    const std::string& imsi, const struct ip_flow_dl& flow_dl,
+    uint32_t flow_precedence, const std::string& apn, uint32_t ue_state,
+    std::function<void(Status, UESessionContextResponse)> callback) {
+  UESessionSet request = UESessionSet();
+  UESessionContextResponse response;
+
+  PipelinedServiceClient& client = get_instance();
+
+  // Set the UE IPv4 address
+  set_ue_ipv4_addr(ue_ipv4_addr, request);
+
+  // Set the UE IPv6 address
+  set_ue_ipv6_addr(ue_ipv6_addr, request);
+
+  // Set the vlan
+  request.set_vlan(vlan);
+
+  // Set the enb IPv4 address
+  set_gnb_ipv4_addr(enb_ipv4_addr, request);
+
+  // Set the incoming and outgoing teid
+  request.set_in_teid(in_teid);
+  request.set_out_teid(out_teid);
+
+  // Set the Subscriber ID
+  SubscriberID* sid = request.mutable_subscriber_id();
+  sid->set_id("IMSI" + std::string(imsi));
+  sid->set_type(SubscriberID::IMSI);
+
+  // Flow dl
+  request.mutable_ip_flow_dl()->CopyFrom(to_proto_ip_flow_dl(flow_dl));
+
+  // Set the precedence
+  request.set_precedence(flow_precedence);
+
+  // Set the APN
+  request.set_apn(apn);
+
+  // Set the ue_state
+  config_ue_session_state(ue_state, request);
+
+  auto local_response = new AsyncLocalResponse<UESessionContextResponse>(
+      std::move(callback), RESPONSE_TIMEOUT);
+
+  auto response_reader = client.stub_->AsyncUpdateUEState(
+      local_response->get_context(), request, &client.queue_);
+
+  local_response->set_response_reader(std::move(response_reader));
+
+  return RETURNok;
+}  // namespace lte
+
+//------------------- TUNNEL DEL -------------------
+
+//                    DEL : v4
+//-------------------------------------------------------------------------
+int PipelinedServiceClient::UpdateUEIPv4SessionSet(
+    struct in_addr& enb_ipv4_addr, const struct in_addr& ue_ipv4_addr,
+    uint32_t in_teid, uint32_t out_teid, uint32_t ue_state,
+    std::function<void(grpc::Status, UESessionContextResponse)> callback) {
+  UESessionSet request = UESessionSet();
+  UESessionContextResponse response;
```

Comment:
Same comment as above about moving the conversion logic into the other file.

Reply:
Comments addressed.